### PR TITLE
[FINERACT-1888] Chargeback acoounting fix

### PR DESCRIPTION
**Pull request by taskain7**
## Description

Fixing accounting for chargeback when the loan was overpaid.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/data/LoanDTO.java [hunk 1]
```diff
@@ -45,4 +46,6 @@ public class LoanDTO {
     private boolean markedAsChargeOff;
     @Setter
     private boolean markedAsFraud;
+    @Setter
+    private BigDecimal overpaidAmount;
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/data/LoanDTO.java]: Not needed, overpayment amount can be found in the `LoanTransactionDTO`

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccountingProcessorHelper.java [hunk 2]
```diff
@@ -786,7 +787,7 @@ public Office getOfficeById(final long officeId) {
         return this.officeRepository.getReferenceById(officeId);
     }
 
-    private void createCreditJournalEntryOrReversalForLoan(final Office office, final String currencyCode, final int accountMappingTypeId,
+    public void createCreditJournalEntryOrReversalForLoan(final Office office, final String currencyCode, final int accountMappingTypeId,
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccountingProcessorHelper.java]: Unnecessary to change the visibility

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccountingProcessorHelper.java [hunk 3]
```diff
@@ -110,6 +110,7 @@ public LoanDTO populateLoanDtoFromMap(final Map<String, Object> accountingBridge
         boolean isAccountTransfer = (Boolean) accountingBridgeData.get("isAccountTransfer");
         boolean isLoanMarkedAsChargeOff = (Boolean) accountingBridgeData.get("isChargeOff");
         boolean isLoanMarkedAsFraud = (Boolean) accountingBridgeData.get("isFraud");
+        BigDecimal overpaidAmount = (BigDecimal) accountingBridgeData.get("overpaidAmount");
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccountingProcessorHelper.java]: You dont need this, LoanTransactionDTO contains the overpaid amount

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccrualBasedAccountingProcessorForLoan.java [hunk 4]
```diff
@@ -418,6 +418,7 @@ private void createJournalEntriesForChargeback(LoanDTO loanDTO, LoanTransactionD
         final Long loanProductId = loanDTO.getLoanProductId();
         final Long loanId = loanDTO.getLoanId();
         final String currencyCode = loanDTO.getCurrencyCode();
+        final BigDecimal overpaidAmount = loanDTO.getOverpaidAmount();
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccrualBasedAccountingProcessorForLoan.java]: final BigDecimal overPaymentAmount = loanTransactionDTO.getOverPayment();

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 5]
```diff
@@ -4492,6 +4492,7 @@ public Map<String, Object> deriveAccountingBridgeData(final String currencyCode,
         accountingBridgeData.put("isAccountTransfer", isAccountTransfer);
         accountingBridgeData.put("isChargeOff", isChargedOff());
         accountingBridgeData.put("isFraud", isFraud());
+        accountingBridgeData.put("overpaidAmount", calculateTotalOverpayment().getAmount().abs());
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: No needed

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please kindly check my comments!
Please include test as well!

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/data/LoanDTO.java (hunk 1 above)
Comment (taskain7) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/data/LoanDTO.java]: removed it

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccountingProcessorHelper.java (hunk 2 above)
Comment (taskain7) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccountingProcessorHelper.java]: reworked it, I'm using the public one with the `AccrualAccountsForLoan` parameter type

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccountingProcessorHelper.java (hunk 3 above)
Comment (taskain7) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccountingProcessorHelper.java]: removed it

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 5 above)
Comment (taskain7) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: removed it

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccrualBasedAccountingProcessorForLoan.java (hunk 4 above)
Comment (taskain7) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/AccrualBasedAccountingProcessorForLoan.java]: chaged it to the suggested one

**Review by adamsaghy — APPROVED** [review]
LGTM

**Pull request merged**